sematic/db/models/external_resource_record.py

```diff
@@ -0,0 +1,168 @@
+# Standard Library
+from typing import Any, Dict, Tuple, Type, Union
+
+# Third-party
+from sqlalchemy import Column, types
+from sqlalchemy.orm import validates
+
+# Sematic
+from sematic.db.models.base import Base
+from sematic.db.models.json_encodable_mixin import JSONEncodableMixin
+from sematic.external_resource import ExternalResource, ResourceState
+from sematic.types.serialization import (
+    type_from_json_encodable,
+    type_to_json_encodable,
+    value_from_json_encodable,
+    value_to_json_encodable,
+)
+
+TypeSerialization = Dict[str, Any]
+
+
+class ExternalResourceRecord(Base, JSONEncodableMixin):
+    """A DB record for an ExternalResource and its history.
+
+    Attributes
+    ----------
+    id:
+        The unique id of the external resource
+    resource_state:
+        The current state of the resource
+    locally_allocated:
+        Whether or not the resource was locally allocated
+    status_message:
+        The most recent status message for the resource
+    last_updated_epoch_seconds:
+        The time that the resource was last updated, expressed as epoch seconds
+    type_serialization:
+        The Sematic serialization of the type. This may be different from the
+        serialization for ExternalResource itself, since the resource will be a
+        subclass of ExternalResource.
+    value_serialization:
+        The Sematic serialization of the object. This will contain extra fields
+        unique to the subclasses of ExternalResource. It also does contain
+        json variants of some of the status fields above, but the duplicated values
+        should match.
+    history_serializations:
+        The Sematic serialization of the objects. Any time the ExternalResource object
+        changes in such a way as to make the instances compare as not equal, a new
+        entry will be added to this list. Element 0 is the most recent, element N
+        the oldest.
+    """
+
+    # Q: Why duplicate data that's already in the json of value_serialization as columns?
+    # A: For two reasons:
+    #    1. It gives us freedom to refactor the dataclass for ExternalResource later
+    #    to move fields around, while allowing the database columns to stay stable.
+    #    2. It allows for more efficient queries on the explicit columns rather than
+    #    requiring json traversal.
```

Comment:
I understand, this is smart. But I will argue YAGNI for now. When we do NI, then we can implement this.

Reply:
discussed offline; the point of the refactoring bullet is that when we *do* need to refactor the dataclass, it will be easier if we can assume that we don't need to do a DB migration as well.